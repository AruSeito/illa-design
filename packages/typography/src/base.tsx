/** @jsxImportSource @emotion/react */
import { Ellipsis, EllipsisBuilder } from "./ellipsis-config"
import * as React from "react"
import { FC, Fragment, MutableRefObject, useContext, useEffect, useRef, useState } from "react"
import {
  applyCopyableContainerSize, applyCopyableIconSize,
  applyExpandLabelCss,
  applyFontColor,
  applyFontContentStyle,
  applyOperationSpan,
} from "./base-style"
import { css } from "@storybook/theming"
import mergedToString, { measureElement } from "./measure-element"
import { BaseProps } from "./interface"
import { Copyable, CopyableBuilder } from "./copyable-config"
import useSize from "react-use/lib/useSize"
import { Tooltip } from "@illa-design/tooltip"
import { ConfigProviderContext, ConfigProviderProps, def } from "@illa-design/config-provider"

function getEllipsis(ellipsis?: boolean | Ellipsis, configProviderProps?: ConfigProviderProps): Ellipsis {
  let originEllipsis: Ellipsis
  if (typeof ellipsis == "boolean" && ellipsis) {
    originEllipsis = new EllipsisBuilder().create()
  } else if ((typeof ellipsis == "boolean" && !ellipsis) || ellipsis == undefined) {
    originEllipsis = new EllipsisBuilder().expandable(false).tooltip(false).create()
  } else {
    originEllipsis = ellipsis
  }
  const locale = configProviderProps?.locale?.typography ?? def.typography
  if (originEllipsis.expandLabel == undefined) {
    originEllipsis.expandLabel = locale["expandLabel"]
  }
  return originEllipsis
}

function getCopyable(copyable?: boolean | Copyable, configProviderProps?: ConfigProviderProps): Copyable {
  // get copyable
  let originCopyable: Copyable
  if (typeof copyable == "boolean" && copyable) {
    originCopyable = new CopyableBuilder().create()
  } else if (typeof copyable == "boolean" && !copyable || copyable == undefined) {
    originCopyable = new CopyableBuilder().copyIcon(null).copiedIcon(null).create()
  } else {
    originCopyable = copyable
  }
  const locale = configProviderProps?.locale?.typography ?? def.typography
  if (originCopyable.copyTooltip == undefined) {
    originCopyable.copyTooltip = locale["copyTooltip"]
  }
  if (originCopyable.copiedToolTip == undefined) {
    originCopyable.copiedToolTip = locale["copiedToolTip"]
  }
  return originCopyable
}

function copyToClipboard(text: string) {
  navigator.clipboard.writeText(text).then()
}

export const Base: FC<BaseProps> = (props) => {

  // get props
  const {
    colorScheme = "blackAlpha",
    ellipsis,
    bold,
    disabled,
    mark,
    underline,
    deleted,
    code,
    copyable,
  } = props

  let configContext = useContext<ConfigProviderProps | undefined>(ConfigProviderContext)

  let originEllipsis = getEllipsis(ellipsis, configContext)
  let originCopyable = getCopyable(copyable, configContext)

  // set expandable state
  const [showExpand, setShowExpand] = useState<boolean>(originEllipsis.expandable)
  const [haveShowExpandSize, setHaveShowExpandSize] = useState<boolean>(false)
  const finalShowExpand = originEllipsis.expandable && showExpand

  const [clipShowText, setClipShowText] = useState("")
  const [copied, setCopied] = useState(false)

  // get ref
  const contentRef = useRef<HTMLDivElement>() as MutableRefObject<HTMLDivElement>
  const operationRef = useRef<HTMLElement>() as MutableRefObject<HTMLElement>

  // apply content
  const contentCss = css`
    ${applyFontColor(colorScheme)};
    ${applyFontContentStyle(bold, mark, underline, deleted, disabled, code)};
  `
  const [content, { width }] = useSize(<span
    ref={contentRef}
    css={contentCss}>
    {finalShowExpand ? clipShowText : props.children}</span>)

  // apply operation
  const copyableElement = <span onClick={() => {
    setCopied(true)
    copyToClipboard(mergedToString(React.Children.toArray(props.children)))
    if (originCopyable.onCopy != undefined) {
      originCopyable.onCopy()
    }
  }} css={applyCopyableContainerSize()}>
    <span css={applyCopyableIconSize}>{!copied ? originCopyable.copyIcon : originCopyable.copiedIcon}</span>
  </span>

  const showCopyTooltip = copied ? originCopyable.copiedToolTip : originCopyable.copyTooltip

  const operation = <span css={applyOperationSpan} ref={operationRef}>
    {finalShowExpand && !haveShowExpandSize && <Fragment>
      <span css={contentCss}>
        ...
        {originEllipsis.suffix && <span css={applyOperationSpan}>{originEllipsis.suffix}</span>}
      </span>
      {<a css={applyExpandLabelCss()} onClick={() => {
        if (originEllipsis.onExpand != undefined) {
          originEllipsis.onExpand()
        }
        setShowExpand(false)
      }}>{originEllipsis.expandLabel}</a>}
    </Fragment>}
    {copyable && originCopyable.copyIcon &&
    showCopyTooltip ?
      <Tooltip closeOnClick={false} content={copied ? originCopyable.copiedToolTip : originCopyable.copyTooltip}>
        {copyableElement}
      </Tooltip> : copyableElement
    }
  </span>

  // update clip text
  useEffect(() => {
    let isMount = true
    if (finalShowExpand) {
      const {
        screenString,
        isClip,
      } = measureElement(contentRef.current, operationRef.current, originEllipsis.rows, props.children)
      if (isMount) {
        setClipShowText(screenString)
        setHaveShowExpandSize(!isClip)
      }
    }
    return () => {
      isMount = false
    }
  }, [width, finalShowExpand])

  return <>
    <span>{content}</span>
    {operation}
  </>
}