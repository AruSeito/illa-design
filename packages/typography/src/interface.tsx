import { HTMLAttributes } from "react"
import { EllipsisConfig } from "./ellipsis-config"

export type TypographyColorScheme =
  "white"
  | "blackAlpha"
  | "gray"
  | "grayBlue"
  | "red"
  | "orange"
  | "yellow"
  | "green"
  | "blue"
  | "cyan"
  | "purple"

export type HeadingLevel = "h1" | "h2" | "h3" | "h4" | "h5" | "h6"

export interface TextProps extends BaseProps, HTMLAttributes<HTMLSpanElement> {
  fontSize: string
}

export interface ParagraphProps extends BaseProps, HTMLAttributes<HTMLParagraphElement> {
  fontSize: string
}

export interface HeadingProps extends BaseProps, HTMLAttributes<HTMLHeadingElement> {
  level?: HeadingLevel
}

export interface BaseProps {
  colorScheme?: TypographyColorScheme
  ellipsis?: boolean | EllipsisConfig
  bold?: boolean
  disabled?: boolean
  mark?: boolean
  underline?: boolean
  deleted?: boolean
  code?: boolean
  copyable?: boolean
}