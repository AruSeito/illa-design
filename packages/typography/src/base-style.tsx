import * as React from "react"
import { TypographyColorScheme } from "./interface"
import { css, SerializedStyles } from "@storybook/theming"
import { globalColor, illaPrefix } from "@illa-design/theme"

const colorSchemes = ["white",
  "blackAlpha",
  "gray",
  "grayBlue",
  "red",
  "orange",
  "yellow",
  "green",
  "blue",
  "cyan",
  "purple"]

export function applyFontColor(colorScheme: TypographyColorScheme): SerializedStyles {

  const color = colorSchemes.includes(colorScheme) ? globalColor(`--${illaPrefix}-${colorScheme}-02`) : colorScheme

  return css`
    color: ${color};
  `
}

export function applyFontContentStyle(bold: boolean, mark: boolean | TypographyColorScheme, underline: boolean, deleted: boolean, disabled: boolean, code: boolean): SerializedStyles {

  let finalCss = css``

  if (code) {
    finalCss = css`
      ${finalCss};
      background-color: ${globalColor(`--${illaPrefix}-grayBlue-05`)};
    `
  }

  if (bold) {
    finalCss = css`
      ${finalCss};
      font-weight: bold;
    `
  }

  if (mark === true) {
    finalCss = css`
      ${finalCss};
      background-color: ${globalColor(`--${illaPrefix}-yellow-06`)};
    `
  } else {
    const bgColor = typeof mark == "string" && colorSchemes.includes(mark) ? globalColor(`--${illaPrefix}-${mark}-06`) : mark
    finalCss = css`
      ${finalCss};
      background-color: ${bgColor};
    `
  }

  finalCss = css`
    ${finalCss};
    text-decoration: ${underline && "underline"} ${deleted && "line-through"};
  `

  if (disabled) {
    finalCss = css`
      ${finalCss};
      cursor: not-allowed;
      color: ${globalColor(`--${illaPrefix}-gray-05`)};
      user-select: none;
    `
  }

  return finalCss
}

export function applyExpandLabelCss(): SerializedStyles {
  return css`
    cursor: pointer;
    margin: 0 4px;
    text-decoration: none;
    color: ${globalColor(`--${illaPrefix}-techBlue-03`)};
  `
}

export function applyCopyableContainerSize(): SerializedStyles {
  return css`
    cursor: pointer;
    display: inline-block;
    margin-left: 4px;
    text-align: center;
    color: ${globalColor(`--${illaPrefix}-gray-02`)};
  `
}