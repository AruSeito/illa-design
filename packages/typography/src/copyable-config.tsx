import { ReactNode } from "react"
import { CopyIcon } from "@illa-design/icon"

export class Copyable {
  copyIcon: ReactNode = <CopyIcon />
  copiedIcon: ReactNode = <CopyIcon />
  tooltips: boolean | string = false
  copyTooltip: boolean | string = false
  onCopy?: () => void
}

export class CopyableBuilder {
  private config: Copyable = new Copyable()

  public copyIcon(copyIcon: ReactNode): CopyableBuilder {
    this.config.copyIcon = copyIcon
    return this
  }

  public copiedIcon(copiedIcon: ReactNode): CopyableBuilder {
    this.config.copiedIcon = copiedIcon
    return this
  }

  public tooltips(tooltips: boolean | string): CopyableBuilder {
    this.config.tooltips = tooltips
    return this
  }

  public copyTooltip(copyTooltip: boolean | string): CopyableBuilder {
    this.config.copyTooltip = copyTooltip
    return this
  }

  public onCopy(onCopy: () => void): CopyableBuilder {
    this.config.onCopy = onCopy
    return this
  }

  public create(): Copyable {
    return this.config
  }
}
