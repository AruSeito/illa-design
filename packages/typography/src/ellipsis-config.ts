import { ReactNode } from "react"

export class EllipsisConfig {
  expandable: boolean = false
  rows: number = -1
  suffix: string = ""
  expandLabel: string | ReactNode = "Expand"
  tooltip: boolean = true
}

export class EllipsisConfigBuilder {
  private config: EllipsisConfig = new EllipsisConfig()

  public expandable(expandable: boolean): EllipsisConfigBuilder {
    this.config.expandable = expandable
    return this
  }

  public rows(rows: number): EllipsisConfigBuilder {
    this.config.rows = rows
    return this
  }

  public suffix(suffix: string): EllipsisConfigBuilder {
    this.config.suffix = suffix
    return this
  }

  public expandLabel(expandLabel: string): EllipsisConfigBuilder {
    this.config.expandLabel = expandLabel
    return this
  }

  public tooltip(tooltip: boolean): EllipsisConfigBuilder {
    this.config.tooltip = tooltip
    return this
  }

  public create(): EllipsisConfig {
    return this.config
  }
}