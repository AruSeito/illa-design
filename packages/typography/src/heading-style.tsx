import * as React from "react"
import { HeadingLevel, TypographyColorScheme } from "./interface"
import { css, SerializedStyles } from "@storybook/theming"
import { globalColor, illaPrefix } from "@illa-design/theme"
import { getHeadingSize } from "./typograph-style"

export function applyFontColor(colorScheme: TypographyColorScheme): SerializedStyles {
  return css`
    color: ${globalColor(`--${illaPrefix}-${colorScheme}-02`)};
  `
}

export function applyContainer(): SerializedStyles {
  return css`
    word-break: break-all;
  `
}

export function applyDisable(disabled: boolean): SerializedStyles {
  if (disabled) {
    return css`
      cursor: not-allowed;
      color: ${globalColor(`--${illaPrefix}-gray-05`)};
      user-select: none;
    `
  } else {
    return css``
  }
}

export function applyFontContentStyle(bold: boolean, mark: boolean, underline: boolean, deleted: boolean, disabled: boolean, code: boolean): SerializedStyles {

  let finalCss = css``

  if (code) {
    finalCss = css`
      ${finalCss};
    `
  }

  if (bold) {
    finalCss = css`
      ${finalCss};
      font-weight: bold;
    `
  }

  if (mark) {
    finalCss = css`
      ${finalCss};
      background-color: ${globalColor(`--${illaPrefix}-yellow-06`)};
    `
  }

  finalCss = css`
    ${finalCss};
    text-decoration: ${underline && "underline"} ${deleted && "line-through"};
  `

  if (disabled) {
    finalCss = css`
      ${finalCss};
      cursor: not-allowed;
      color: ${globalColor(`--${illaPrefix}-gray-05`)};
      user-select: none;
    `
  }

  return finalCss
}

export function applyExpandLabelCss(): SerializedStyles {
  return css`
    cursor: pointer;
    color: ${globalColor(`--${illaPrefix}--techBlue-03`)};
  `
}

export function applyCopyableContainerSize(level: HeadingLevel): SerializedStyles {
  return css`
    cursor: pointer;
    color: ${globalColor(`--${illaPrefix}--gray-02`)};
    line-height: ${getHeadingSize(level)[1]};
    height: ${getHeadingSize(level)[1]};
    display: inline-block;
    margin-left: 8px;
    text-align: center;
  `
}