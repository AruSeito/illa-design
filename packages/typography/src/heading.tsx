/** @jsxImportSource @emotion/react */
import * as React from "react"
import { forwardRef } from "react"
import { HeadingProps } from "./interface"
import { Base } from "./base"
import { Tooltip } from "@illa-design/tooltip"
import mergedToString from "./measure-element"

export const Heading = forwardRef<HTMLHeadingElement, HeadingProps>((props, ref) => {

  // get props
  const {
    colorScheme = "blackAlpha",
    level = "h2",
    ellipsis,
    bold,
    disabled,
    mark,
    underline,
    deleted,
    code,
    copyable,
    ...otherProps
  } = props

  const showTooltip = !disabled && (ellipsis == true || (typeof ellipsis == "object" && ellipsis.tooltip))

  const base = <Base colorScheme={colorScheme} ellipsis={ellipsis} bold={bold} disabled={disabled} mark={mark}
                     underline={underline} deleted={deleted} code={code} copyable={copyable}>
    {props.children}
  </Base>

  let headingNode

  switch (level) {
    case "h1":
      headingNode = <h1 ref={ref} {...otherProps}>
        {base}
      </h1>
      break
    case "h2":
      headingNode = <h2 ref={ref} {...otherProps}>
        {base}
      </h2>
      break
    case "h3":
      headingNode = <h3 ref={ref} {...otherProps}>
        {base}
      </h3>
      break
    case "h4":
      headingNode = <h4 ref={ref} {...otherProps}>
        {base}
      </h4>
      break
    case "h5":
      headingNode = <h5 ref={ref} {...otherProps}>
        {base}
      </h5>
      break
    case "h6":
      headingNode = <h6 ref={ref} {...otherProps}>
        {base}
      </h6>
      break
  }
  if (showTooltip) {
    return <Tooltip content={mergedToString(React.Children.toArray(props.children))}>
      {headingNode}
    </Tooltip>
  } else {
    return headingNode
  }
})
