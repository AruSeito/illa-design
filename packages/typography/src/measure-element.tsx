import * as React from "react"
import { unmountComponentAtNode } from "react-dom"
import { leftIcon } from "@illa-design/tag/dist/types/style"

let computeElement: HTMLElement

function inRange(maxHeight: number, lastLineMaxWidth: number, rows: number) {
  const lines = computeElement.getClientRects().length
  if (lines > rows) {
    return false
  }
  if (lines < rows) {
    return true
  }
  const lastLineWidth = computeElement.getClientRects().item(rows - 1)?.width ?? 0
  return lastLineWidth <= lastLineMaxWidth
}

function measureText(textNode: Text, fullText: string, maxHeight: number, lastLineMaxWidth: number, rows: number, start: number = 0, end: number = fullText.length) {
  if (start >= end - 1 && inRange(maxHeight, lastLineMaxWidth, rows)) {
    return
  }
  const mid = (start + end) / 2
  textNode.textContent = fullText.slice(0, mid)
  if (inRange(maxHeight, lastLineMaxWidth, rows)) {
    measureText(textNode, fullText, maxHeight, lastLineMaxWidth, rows, mid, end)
  } else {
    measureText(textNode, fullText, maxHeight, lastLineMaxWidth, rows, start, mid)
  }
}

function getContentHeight(contentRef: HTMLSpanElement): number {
  let height = 0
  for (let c of contentRef.getClientRects()) {
    height += c.height
  }
  return height
}

function getContentWidth(contentRef: HTMLSpanElement): number {
  let width = 0
  for (let c of contentRef.getClientRects()) {
    width += c.width
  }
  return width
}

function getMaxLineWidth(contentRef: HTMLSpanElement): number {
  let maxWidth = 0
  for (let c of contentRef.getClientRects()) {
    maxWidth = Math.max(maxWidth, c.width)
  }
  return maxWidth
}

export function measureElement(
  contentRef: HTMLElement,
  operationRef: HTMLSpanElement,
  rows: number,
): [string, boolean] {
  const lineHeight = contentRef.getClientRects().item(0)?.height ?? 0
  const operationWidth = getContentWidth(operationRef)
  const lastLineMaxWidth = getMaxLineWidth(contentRef) - operationWidth

  if (computeElement == undefined) {
    computeElement = document.createElement(contentRef.tagName)
    document.body.appendChild(computeElement)
  }

  // create text node
  const fullText = contentRef.textContent ?? ""
  const textNode = document.createTextNode(fullText)

  // deal css
  const originStyle = window.getComputedStyle(contentRef)
  const styleNames: string[] = Array.prototype.slice.apply(originStyle)
  const styleString = styleNames
    .map((name) => `${name}: ${originStyle.getPropertyValue(name)};`)
    .join("")

  computeElement.setAttribute("style", styleString)
  computeElement.insertBefore(textNode, computeElement.firstChild)

  const maxHeight = lineHeight * rows

  if (getContentHeight(contentRef) <= maxHeight) {
    unmountComponentAtNode(computeElement)
    computeElement.innerText = ""
    return [fullText, false]
  }

  measureText(textNode, fullText, maxHeight, lastLineMaxWidth, rows)
  const finalString = computeElement.textContent ?? ""
  unmountComponentAtNode(computeElement)
  computeElement.innerText = ""

  return [finalString, true]
}
