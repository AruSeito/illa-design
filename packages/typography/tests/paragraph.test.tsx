import * as React from "react"
import { fireEvent, render, screen } from "@testing-library/react"
import "@testing-library/jest-dom"
import { EllipsisBuilder, Paragraph, Typography } from "../src"

test("Paragraph renders with different font size", () => {
  render(<Typography>
    <Paragraph data-testid="test-paragraph" fontSize="20px">A design is a plan or specification for the construction of
      an object or system or for
      the
      implementation of an activity or process, or the result of that plan or specification in the form of a prototype,
      product or process. The verb to design expresses the process of developing a design. In some cases, the direct
      construction of an object without an explicit prior plan (such as in craft work, some engineering, coding, and
      graphic design) may also be considered to be a design activity.</Paragraph>
  </Typography>)
  expect(screen.getByTestId("test-paragraph")).toBeInTheDocument()
})

test("Paragraph renders with indent", () => {
  render(<Typography>
    <Paragraph data-testid="test-paragraph" indent={true} fontSize="20px">A design is a plan or specification for the
      construction of
      an object or system or for
      the
      implementation of an activity or process, or the result of that plan or specification in the form of a prototype,
      product or process. The verb to design expresses the process of developing a design. In some cases, the direct
      construction of an object without an explicit prior plan (such as in craft work, some engineering, coding, and
      graphic design) may also be considered to be a design activity.</Paragraph>
  </Typography>)
  expect(screen.getByTestId("test-paragraph")).toHaveStyle({
    "text-indent": "32px",
  })
})

test("Paragraph renders with styles", () => {
  render(<Typography>
    <Paragraph data-testid="test-paragraph" underline={true} mark={true} deleted={true} bold={true}>A design is a plan
      or
      specification for the
      construction of
      an object or system or for
      the
      implementation of an activity or process, or the result of that plan or specification in the form of a prototype,
      product or process. The verb to design expresses the process of developing a design. In some cases, the direct
      construction of an object without an explicit prior plan (such as in craf twork, some engineering, coding, and
      graphic design) may also be considered to be a design activity.</Paragraph>
  </Typography>)
  expect(screen.getByTestId("test-paragraph")).toMatchSnapshot()
})

test("Paragraph renders with code style", () => {
  render(<Typography>
    <Paragraph data-testid="test-paragraph" code={true} ellipsis={false}>A design is a plan or
      specification for the
      construction of
      an object or system or for
      the
      implementation of an activity or process, or the result of that plan or specification in the form of a prototype,
      product or process. The verb to design expresses the process of developing a design. In some cases, the direct
      construction of an object without an explicit prior plan (such as in craft work, some engineering, coding, and
      graphic design) may also be considered to be a design activity.</Paragraph>
  </Typography>)
  expect(screen.getByTestId("test-paragraph")).toMatchSnapshot()
})

test("Paragraph renders with disabled style", () => {
  render(<Typography>
    <Paragraph data-testid="test-paragraph" disabled={true}>
      This is a paragraph
    </Paragraph>
  </Typography>)
  expect(screen.getByText("This is a paragraph")).toMatchSnapshot()
  fireEvent.mouseEnter(screen.getByText("This is a paragraph"))
  expect(screen.getByText("This is a paragraph")).toHaveStyle({
    "cursor": "not-allowed",
  })
})

test("Paragraph renders short text with ellipsis", () => {
  render(<Typography>
    <Paragraph ellipsis={true}>
      This is a paragraph
    </Paragraph>
  </Typography>)
  expect(screen.queryByText("...")).not.toBeInTheDocument()
})

test("Paragraph render with ellipsis", () => {
  const onExpand = jest.fn()
  render(<Typography>
    <Paragraph data-testid="test-paragraph"
               ellipsis={new EllipsisBuilder()
                 .rows(2)
                 .expandable(true)
                 .expandLabel("TestExpand")
                 .onExpand(onExpand)
                 .suffix("TestSuffix")
                 .create()}>
      A design is a plan or
      specification for the
      construction of
      an object or system or for
      the
      implementation of an activity or process, or the result of that plan or specification in the form of a prototype,
      product or process. The verb to design expresses the process of developing a design. In some cases, the direct
      construction of an object without an explicit prior plan (such as in craft work, some engineering, coding, and
      graphic design) may also be considered to be a design activity.
    </Paragraph>
  </Typography>)
  expect(screen.getByText("TestSuffix")).toBeInTheDocument()
  expect(screen.getByText("...")).toBeInTheDocument()
  expect(screen.getByText("TestExpand")).toBeInTheDocument()
  fireEvent.click(screen.getByText("TestExpand"))
  expect(onExpand).toBeCalled()
  expect(screen.queryByText("...")).not.toBeInTheDocument()
})

