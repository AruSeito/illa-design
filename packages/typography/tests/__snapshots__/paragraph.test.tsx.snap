// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paragraph renders with code style 1`] = `
.emotion-1 {
  color: rgba(0, 0, 0, 0.64);
  background-color: #f3f5f7;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-3 {
  cursor: pointer;
  display: inline-block;
  margin-left: 4px;
  text-align: center;
  color: #1f1f1f;
}

<div
  class="emotion-0"
  data-testid="test-paragraph"
>
  <span>
    <span
      class="emotion-1"
      style="position: relative;"
    >
      <iframe
        style="background: transparent; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%; z-index: -1;"
      />
      A design is a plan or specification for the construction of an object or system or for the implementation of an activity or process, or the result of that plan or specification in the form of a prototype, product or process. The verb to design expresses the process of developing a design. In some cases, the direct construction of an object without an explicit prior plan (such as in craft work, some engineering, coding, and graphic design) may also be considered to be a design activity.
    </span>
  </span>
  <span
    class="emotion-2"
  >
    <span
      class="emotion-3"
    />
  </span>
</div>
`;

exports[`Paragraph renders with disabled style 1`] = `
.emotion-0 {
  color: rgba(0, 0, 0, 0.64);
  cursor: not-allowed;
  color: #c2c2c2;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

<span
  class="emotion-0"
  style="position: relative;"
>
  <iframe
    style="background: transparent; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%; z-index: -1;"
  />
  This is a paragraph
</span>
`;

exports[`Paragraph renders with styles 1`] = `
.emotion-1 {
  color: rgba(0, 0, 0, 0.64);
  font-weight: bold;
  background-color: #fff2a3;
  -webkit-text-decoration: underline line-through;
  text-decoration: underline line-through;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-3 {
  cursor: pointer;
  display: inline-block;
  margin-left: 4px;
  text-align: center;
  color: #1f1f1f;
}

<div
  class="emotion-0"
  data-testid="test-paragraph"
>
  <span>
    <span
      class="emotion-1"
      style="position: relative;"
    >
      <iframe
        style="background: transparent; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%; z-index: -1;"
      />
      A design is a plan or specification for the construction of an object or system or for the implementation of an activity or process, or the result of that plan or specification in the form of a prototype, product or process. The verb to design expresses the process of developing a design. In some cases, the direct construction of an object without an explicit prior plan (such as in craft work, some engineering, coding, and graphic design) may also be considered to be a design activity.
    </span>
  </span>
  <span
    class="emotion-2"
  >
    <span
      class="emotion-3"
    />
  </span>
</div>
`;
